lib/elixir/lib/keyword.ex

```
@@ -186,6 +186,193 @@ defmodule Keyword do
+  end
+
+  @doc """
+  Ensures the first argument is a `keyword` with the given
+  keys and default values.
+  The second argument must be a list of atoms, specifying
+  a given key, or tuples specifying a key and a default value.
+
+  Possible return values:
+  * `{:ok, keyword}`: Success case. Returns the validated keyword with default values applied.
+  * `{:error, :not_a_keyword}`: Indicates that the validated value is not a keyword.
+  * `{:error, :invalid_expected_values}`: Indicates that the expected values list is invalid.
+  * `{:error, {:invalid_key, key}}`: Indicates that a given key is not expected.
+
+  See also: `validate!/2`
+
+  ## Examples
+      iex> {:ok, result} = Keyword.validate([], [one: 1, two: 2])
+      iex> Enum.sort(result)
+      [one: 1, two: 2]
+
+      iex> {:ok, result} = Keyword.validate([two: 3], [one: 1, two: 2])
+      iex> Enum.sort(result)
+      [one: 1, two: 3]
+
+  If atoms are given, they are supported as keys but do not
+  provide a default value:
+
+      iex> {:ok, result} = Keyword.validate([], [:one, two: 2])
+      iex> Enum.sort(result)
+      [two: 2]
+
+      iex> {:ok, result} = Keyword.validate([one: 1], [:one, two: 2])
+      iex> Enum.sort(result)
+      [one: 1, two: 2]
+
+  Passing an unknown key returns an error:
+
+      iex> Keyword.validate([three: 3], [one: 1, two: 2])
+      {:error, {:invalid_key, :three}}
+
+  Passing an invalid value in either argument also returns errors
+
+      iex> Keyword.validate(%{three: 3}, [one: 1, two: 2])
+      {:error, :not_a_keyword}
+      iex> Keyword.validate([:three], [one: 1, two: 2])
+      {:error, :not_a_keyword}
+      iex> Keyword.validate([three: 3], [{"one", 1}, :two])
+      {:error, :invalid_expected_values}
+  """
+  @spec validate(keyword :: keyword(), values :: [atom() | {atom(), term()}]) ::
+          {:ok, keyword}
+          | {:error,
+             :not_a_keyword
+             | {:invalid_key, term()}
+             | :invalid_expected_values}
```

per the ML thread:

```suggestion
          {:ok, keyword} | {:error, unknown_keys, missing_keys}
```

---

To add to @wojtekmach's comment, invalid inputs (i.e. not a keyword or invalid expected values) should just raise.

---

Actually, my comment on the mailing list was wrong, there is no such thing as missing keys as we don't have required keys. So this should simply be:

    {:ok, keyword} | {:error, unknown_keys :: [atom]}

---

Oh, good call. In that case, `validate!` should definitely raise a KeyError which will give us a really good error message.

---

scratch that, we can't raise a KeyError as we could have multiple unknown keys.

---

@josevalim I've updated the code with the reviews. One thing that isn't clear to me is that the suggested spec shows a list of invalid keys.

The current implementation only returns the first invalid key. Should I change this for `validate`?